### PR TITLE
NAS-118493 / 22.12 / Add integration test for testing port delegates

### PR DESCRIPTION
**Pull request by sonicaj**

### CODE REVIEW

File: tests/api2/test_port_delegates.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+def test_port_delegate_validation_with_invalid_ports():
+    in_use_ports = []
+    for entry in call('port.get_in_use'):
+        in_use_ports.extend(filter(lambda i: i > 1024, entry['ports']))
+
+    for config_method, method, keys, payload in PAYLOAD:
+        validation_error = None
+        for key in keys:
+            payload[key] = in_use_ports[random.randint(0, len(in_use_ports) - 1)]
```
Comment (themylogin) [inline tests/api2/test_port_delegates.py:28]: Why do you use `random`? It's a recipe to non-repeatable failures or false negatives

**Review by themylogin — CHANGES_REQUESTED** [review]

File: tests/api2/test_port_delegates.py (hunk 1 above)
Comment (sonicaj) [reply tests/api2/test_port_delegates.py:28]: We should be choosing any port which is already being consumed so i wanted to randomly select any element in the list. Should we remove it instead ? ( btw wrt this usage i think we should be safe )

File: tests/api2/test_port_delegates.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def test_port_delegate_validation_with_invalid_ports():
+    in_use_ports = []
+    for entry in call('port.get_in_use'):
+        in_use_ports.extend(filter(lambda i: i > 1024, entry['ports']))
+
+    assert in_use_ports != [], 'No in use ports retrieved'
+
+    for config_method, method, keys, payload in PAYLOAD:
```
Comment (themylogin) [inline tests/api2/test_port_delegates.py:28]: It's better to use `pytest.mark.parametrize` and try each payload in a separate test. If you use `for` loop, one payload failure will result to subsequent not being tested.

**Review by themylogin — CHANGES_REQUESTED** [review]

File: tests/api2/test_port_delegates.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    validation_error = None
+    for index, key in enumerate(keys):
+        payload[key] = in_use_ports[index] if len(in_use_ports) > index else in_use_ports[0]
+    try:
+        call(method, payload)
+    except ValidationErrors as ve:
+        validation_error = ve
+
+    assert validation_error is not None, 'Port validation exception expected'
```
Comment (themylogin) [inline tests/api2/test_port_delegates.py:38]: Sorry, one more thing. To ensure that an exception is raised use this:
```
    with pytest.raises(ValidationErrors) as ve:
        call("privilege.update", 1, {"allowlist": [{"method": "CALL", "resource": "system.info"}]})

    assert ve.value.errors[0].attribute == "privilege_update.allowlist"
```

**Review by themylogin — APPROVED** [review]
LGTM, thank you very much!

**Pull request merged**